pyomo/contrib/pyros/uncertainty_sets.py

```diff
@@ -1,99 +1,378 @@
 """
-UncertaintySet class: defines generic methods and attributes
-of an uncertainty set in the context of robust optimization. UncertaintySet objects only
-contain data which describes the set, and does not contain any Pyomo object information.
-
-Supports the following classes of uncertainty sets:
-
-- UncertaintySet (user defined/implemented)
-- Ellipsoidal
-- AxesAlignedEllipsoidal
-- Polyhedral
-- Box
-- BudgetSet
-- Cardinality/Gamma
-- Discrete
-- FactorModel
-- IntersectedSet
+Abstract and pre-defined classes for representing uncertainty sets (or
+uncertain parameter spaces) of two-stage nonlinear robust optimization
+models.
+
+Along with a ``ConcreteModel`` object representing a deterministic model
+formulation, an uncertainty set object may be passed to the PyROS solver
+to obtain a solution to the model's two-stage robust optimization
+counterpart.
+
+Classes
+-------
+``UncertaintySet``
+    Abstract base class for a generic uncertainty set. All other set
+    types defined in this module are subclasses.  A user may implement
+    their own uncertainty set type as a custom-written subclass.
+
+``EllipsoidalSet``
+    A hyperellipsoid.
+
+``AxisAlignedEllipsoidalSet``
+    An axis-aligned hyperellipsoid.
+
+``PolyhedralSet``
+    A bounded convex polyhedron/polytope.
+
+``BoxSet``
+    A hyperrectangle.
+
+``BudgetSet``
+    A budget set.
+
+``CardinalitySet``
+    A cardinality set (or gamma set).
+
+``DiscreteScenarioSet``
+    A discrete set of finitely many points.
+
+``FactorModelSet``
+    A factor model set (or net-alpha model set).
+
+``IntersectionSet``
+    An intersection of two or more sets, each represented by an
+    ``UncertaintySet`` object.
 """
 
+
 import abc
-import functools
 import math
+import functools
+from numbers import Integral
+from collections.abc import Iterable, MutableSequence
 from enum import Enum
+
 from pyomo.common.dependencies import numpy as np, scipy as sp
 from pyomo.core.base import ConcreteModel, Objective, maximize, minimize, Block
 from pyomo.core.base.constraint import ConstraintList
-from pyomo.core.base.var import Var, _VarData, IndexedVar
-from pyomo.core.base.param import Param, _ParamData, IndexedParam
-from pyomo.core.expr import value
+from pyomo.core.base.var import Var, IndexedVar
+from pyomo.core.expr.numvalue import value, native_numeric_types
 from pyomo.opt.results import check_optimal_termination
 from pyomo.contrib.pyros.util import add_bounds_for_uncertain_parameters
 
+
+valid_num_types = tuple(native_numeric_types)
```

Comment:
This is OK for now, but I would recommend adding an issue to update this in the future: the `native_numeric_types` set is mutable for a reason: it may get additional "numeric-like" types added to it when they are encountered in a user's model (e.g, numpy types).  By making a copy, you are freezing the set whenever pyros gets imported.

You might consider accepting `set` to `validate_arg` and then convert it to a tuple before using it in `isinstance`.

Further, converting the set to a tuple has a negative impact on the performance of both `issubset` and `in`.

Reply:
Noted.